apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileThread.java

```diff
@@ -0,0 +1,296 @@
+                        slot.setStartDump(true);
+                    }
+
+                }
+            }
+
+            // sleep to next period
+            // if out of period, sleep one period
+            long needToSleep = (currentLoopStartTime + maxSleepPeriod) - System.currentTimeMillis();
+            needToSleep = needToSleep > 0 ? needToSleep : maxSleepPeriod;
+            Thread.sleep(needToSleep);
+        }
+    }
+
+    /**
+     * dump segemnt thread stack
+     * @param segmentContext
+     * @return
+     */
+    private boolean dumpSegment(ProfilingSegmentContext segmentContext) {
+        // dump stack
+        if (!checkSegmentProfilingCanContinue(segmentContext)) {
+            return false;
+        }
+
+        return dumpThread(segmentContext);
+    }
+
+    /**
+     * dump thread stack, and push data to backend
+     * @param segmentContext
+     * @return still can dump
+     */
+    private boolean dumpThread(ProfilingSegmentContext segmentContext) {
+        long currentTime = System.currentTimeMillis();
+        // dump thread
+        final StackTraceElement[] stackTrace = segmentContext.getProfilingThread().getStackTrace();
+
+        // stack depth is zero, means thread is already run finished
+        if (stackTrace.length == 0) {
+            return false;
+        }
+
+        int dumpElementCount = Math.min(stackTrace.length, Config.Profile.DUMP_MAX_STACK_DEPTH);
+
+        // use inverted order, because thread dump is start with bottom
+        final ArrayList<String> stackList = new ArrayList<>(dumpElementCount);
+        for (int i = dumpElementCount - 1; i >= 0; i--) {
+            stackList.add(buildStackElementCodeSignature(stackTrace[i]));
+        }
+
+        // build snapshot and send
+        ProfileTaskSegmentSnapshot snapshot = new ProfileTaskSegmentSnapshot(segmentContext, segmentContext.getCurrentAndIncrementSequence(), currentTime, stackList);
+        profileTaskChannelService.addProfilingSnapshot(snapshot);
+        return true;
+    }
+
+    private String buildStackElementCodeSignature(StackTraceElement element) {
+        // className.methodName:lineNumber
+        return element.getClassName() + "." + element.getMethodName() + ":" + element.getLineNumber();
+    }
+
+    /**
+     * check segment profiling is should continue
+     * @param context
+     * @return
+     */
+    private boolean checkSegmentProfilingCanContinue(ProfilingSegmentContext context) {
+        // check segment still executing
+        if (!context.getSegmentIsRunning()) {
+            return false;
+        }
+
+        // check is out of limit monitor time
+        if (System.currentTimeMillis() - context.getProfilingStartTime() > MAX_PROFILING_TIME_MILLS) {
+            return false;
+        }
+
+        // check segment executing thread is still running
+        if (!context.getProfilingThread().isAlive()) {
```

Comment:
Don't check this every time, once you don't get the notification event(`notifyAfterMainThreadFinish`), it is safe to continue.